cred/src/main/java/org/opencadc/cred/GetCertAction.java

```diff
@@ -1,96 +1,310 @@
- * 
- * @author pdowler
+ * Class to handle certificate generation requests
+ * @author adriand
  */
 public class GetCertAction extends RestAction {
-    private static final Logger log = Logger.getLogger(GetCertAction.class);
+    protected static Logger log = Logger.getLogger(GetCertAction.class);
+
+    static final String CERTIFICATE_CONTENT_TYPE = "application/x-pem-file";
+    static final String CERTIFICATE_FILENAME = "cadcproxy.pem"; // content disposition
+    static final int CERT_KEY_SIZE = 2048;
+
+    // CADC specific fields of the DN
+    public static final String CADC_DN = "ou=cadc,o=hia,c=ca";
+
+    private CredConfig config;
 
-    public GetCertAction() { 
+    public GetCertAction() {
+        super();
     }
 
     @Override
-    protected InlineContentHandler getInlineContentHandler() {
+    protected final InlineContentHandler getInlineContentHandler() {
         return null;
     }
 
+    @Override
+    public void initAction() {
+        config = CredInitAction.getConfig(super.appName);
+    }
+
     @Override
     public void doAction() throws Exception {
-        throw new UnsupportedOperationException();
+        // create a cert for a single user
+        Subject caller = AuthenticationUtil.getCurrentSubject();
+        String path = syncInput.getPath();
+        if (AuthMethod.CERT.equals(AuthenticationUtil.getAuthMethod(caller)) && !StringUtil.hasText(path)) {
+            throw new AccessControlException("Cert Authentication not allowed for cert renewal.");
+        }
+        Set<X500Principal> dnPrincipals = caller.getPrincipals(X500Principal.class);
+        if (dnPrincipals.size() != 1) {
+            throw new NotAuthenticatedException("Authentication required (caller DN not found).");
+        }
+        X500Principal callerDN = dnPrincipals.iterator().next();
+
+
+        X500Principal userDN = callerDN;
+        if (StringUtil.hasText(path)) {
+            path = path.replace("^/+", "").replace("/+$", "");
+            log.debug("User ID path " + path);
+            Principal delegatedUser = getPrincipal(path);
+            if (config.superUsers.contains(callerDN)) {
+                if (delegatedUser instanceof X500Principal) {
+                    // no need for augment subject
+                    userDN = (X500Principal)delegatedUser;
+                } else {
+                    Subject delegatedSub = new Subject();
+                    delegatedSub.getPrincipals().add(delegatedUser);
+                    AuthenticationUtil.augmentSubject(delegatedSub);
+                    dnPrincipals = delegatedSub.getPrincipals(X500Principal.class);
+                    if (dnPrincipals.size() != 1) {
+                        throw new NotAuthenticatedException("User not found: " + delegatedUser);
+                    }
+                    userDN = dnPrincipals.iterator().next();
+                }
+            } else {
+                throw new AccessControlException("Not a superuser: " + callerDN);
```

Comment:
This would probably catch the first issue but the error would be wrong and unhelpful. Instead, there should be an explicit 
```
if (caller == user) {
    throw new AccessControlException("cannot renew a temporary certificate with a previously issued certificate");
}
```
It should be possible for me to use my real grid canada cert to get an internal (operator-signed) cert, but I don't know if it's feasible to allow for that. You'd have to compare the issuer of the caller with the signing cert subject?